### PR TITLE
fix(#933): a reviewer quoting the author's SHA no longer proves it read HEAD

### LINKED ISSUE #933: review-substance.sh: status_comment_names_head can be satisfied by a bot echoing the author's own SHA-bearing comment
Found during PR #929's Phase B (Sat Aug 1 2026).

## Observation

`review-substance.sh` set `status_comment_names_head=true` for a CodeAnt re-review partly because the bot's comment body contained the HEAD short SHA `5acd1e2` — but that string had been quoted into the thread by the PR author's own re-review trigger prose. A bot that merely **echoes or quotes the author's comment** (common in ack/summary replies) could therefore flip the flag with no genuine review behind it.

On PR #929 this was harmless — the review also enumerated round-1 edits that exist only at HEAD, and an independent BugBot clean pass named the full SHA — but the flag's evidentiary value is weaker than designed when the SHA string can originate from the author.

## Suggested direction

- When scanning a bot's status/summary comment for the HEAD SHA, discount matches that appear inside quoted/echoed author text (e.g. blockquote lines, `>`-prefixed content, or verbatim substrings of recent author comments on the same thread).
- Alternatively (cheaper): require the SHA mention to co-occur with the bot's own run-marker phrasing (e.g. its "finished running"/"Reviewed commit" template) rather than counting any body containing the string.
- Regression fixtures: (a) bot comment quoting the author's SHA-bearing trigger → must NOT count; (b) genuine run-marker comment naming HEAD → counts (parity with today).

## Context

- Guard lineage: issue #875 / PR #883 (hollow approvals), issue #876 / PR #893 (in-place PATCH re-reviews), issue #894 / PR #896 (all-decimal short SHA).
- Related memory: post-push-bot-rereview-tells (a stale timestamp is redeemed by external evidence on HEAD, never by reviewer identity).

_Filed by the PM session that orchestrated PR #929._

### PR DESCRIPTION
**Pull request by auerbachb**
## **User description**
Closes #933

## What was wrong

`status_comment_names_head` is the flag that asserts *this reviewer demonstrably
read HEAD*. It was set by any conversation comment from that bot containing
HEAD's SHA — including a SHA the bot never wrote.

On PR #929, CodeAnt answers a re-review request by reproducing the requester's
comment **verbatim** under a `Question:` heading, with its own verdict under
`Answer:`. The author's trigger prose named `5acd1e2`, so the flag went true on
the **author's** words and an empty-bodied approval scored as full review
coverage.

## What changed

`sha_tokens` drops, before extraction, any line already posted on the thread by
someone who is **not** one of the reviewers under evaluation. Quote markers are
normalised out of the comparison key, so GitHub's "Quote reply" shape matches its
source. `merge-gate.sh` already passes the unfiltered issue-comments list, so
nothing new crosses the `REVIEW_EVIDENCE` boundary.

**Not "strip every blockquote line"** — that was the first cut, and the live
payload refutes it twice. The #929 echo carries no `>` at all, so a blockquote
rule closes *none* of the reported trace; and the only `>` lines on that thread
are CodeRabbit quoting **itself** (`> Reviewing files that changed … between X
and Y`). Blanket quote-stripping would have deleted a reviewer's own self-report
on every CodeRabbit status comment in this repo while fixing nothing. The issue
asks to discount quoted/echoed *author* text; a bot's own callout is not that.

Guards against over-reach: reviewers never seed the index (so a reviewer can
never strip itself, including one repeating its own status line); the index
stores the earliest time each line was seen (so a human quoting a bot cannot
retroactively strip the bot's original); and fence delimiters are exempt, because
dropping a bare ` ``` ` line would break the #897 paired-delimiter masking.

### Directions — one of them is a grant

- `status_comment_names_head`, `external_evidence_on_head`, `substantive` are
  monotone in the token set → **true → false only**. This is the grant path #933
  closes.
- `self_report_mismatch` moves **both** ways, and `counts_as_coverage` inherits
  that single upward channel: a mismatch manufactured purely from quoted text is
  retracted. That is the same correction, not a side effect — a SHA the reviewer
  never wrote was never its report about itself — and it is the direction #917
  already accepted for UUID fragments.

An earlier draft of the rationale claimed `counts_as_coverage` was withhold-only.
The CodeRabbit CLI caught that overclaim; it is corrected in both the script and
`review-substance-evidence.md`.

## Evidence

**Live replay** (main's evaluator vs this branch, on the real PR #929 payload at
`5acd1e2`), comparing decision flags:

| reviewer | main | branch |
|---|---|---|
| `codeant-ai[bot]` | `counts_as_coverage: true`, shas `["5acd1e2","c441771"]` | `counts_as_coverage: false`, disqualified `no_substantive_footprint` + `self_report_mismatch`, shas `["c441771", …]` |
| `coderabbitai[bot]` | — | **byte-identical**, flags and reporting alike |

`corroborating: ["cursor[bot]"]` unchanged. The branch's `status_comment_shas`
for CodeAnt now reports the SHA its own status comment actually names.

**Non-vacuity control** — the branch's test file run against **main's**
evaluator: `PASS: 136  FAIL: 13`. All 13 failures are the direction cases; every
parity/ordering/self-quotation/fence control passes on both, and there are zero
non-`ff-933` failures.

**Suites:** 149/149 in `merge-gate-review-substance.test.sh`; all 9 suites that
touch this evaluator green; the full CI discovery (`run-hook-tests.sh`) runs
65 suites, 0 failed; `rule-lint.sh` OK; `shellcheck` clean on both changed
scripts, matching main's clean baseline.

**Performance:** amplified to 200 conversation comments (40× the live payload),
243 ms/run on main vs 325 ms/run here. Negligible at real payload sizes and far
from the O(n²) pathology this file's header warns about.

**Known bound (deliberate):** matching is whole-line. A bot that paraphrases or
re-wraps the author's text instead of echoing a full line keeps the token — the
conservative direction, identical to today's behaviour.

**Local review coverage:** none — CodeAnt returned a 403 on both permitted
attempts (Issue #643: persistent account-level entitlement, not auth; `codeant
logout` deliberately not run). CodeRabbit completed two runs and **all 3 of its
findings are fixed in this commit** (MD038 code-span, the `counts_as_coverage`
direction overclaim, and the fence-delimiter interaction — the last shipped with
its own regression case, `(ff-933)(j)`, verified to fail before the guard). Its
verification pass then hung past the 2-minute bound and was dropped without a
clean pass, so a self-review stands in: 6 edge-case probes (empty conversation,
null body, absent user object, absent timestamps, CRLF-vs-LF, trailing
whitespace, indented echo, partial quote) all behave correctly with no evaluator
errors.

## Test plan

- [x] A bot comment whose only HEAD mention is a verbatim echo of the author's
      SHA-bearing trigger does **not** set `status_comment_names_head`, sets no
      `external_evidence_on_head`, and does not count as coverage — `(ff-933)(a)`
- [x] A genuine status comment naming HEAD in the bot's own prose still counts,
      unchanged from today — `(ff-933)(b)`
- [x] A body containing both an echoed line and the bot's own HEAD-naming prose
      still counts: the strip removes the line, not the body — `(ff-933)(c)`
- [x] GitHub's `>`-prefixed "Quote reply" of the author's line is caught too —
      `(ff-933)(d)`
- [x] A bot's own blockquoted callout naming HEAD is **not** treated as an echo
      (negative control for the rejected blanket-blockquote rule) — `(ff-933)(e)`
- [x] An identical author line posted **after** the bot cannot retroactively
      strip the bot's line — `(ff-933)(f)`
- [x] A quoted **older** SHA yields no tokens and manufactures no
      `self_report_mismatch` — `(ff-933)(g)`
- [x] A reviewer repeating its own status line does not erase its own evidence —
      `(ff-933)(h)`
- [x] One reviewer quoting another stays out of scope (boundary pinned) —
      `(ff-933)(i)`
- [x] An echoed fence delimiter still masks its block, so quoted source code is
      not a self-report — `(ff-933)(j)`
- [x] An echoed fence line does **not** smuggle a SHA through its info string:
      the exemption truncates to the delimiter rather than keeping the line, on
      both spellings (` ```<sha> ` and ` ``` <prose with sha> `) — `(ff-933)(k)`
- [x] …and the truncated opener still pairs with its closer, so the block is
      still masked and quoted source is still not a self-report — `(ff-933)(l)`
- [x] An echo **edited in place** into a bot comment that predates the author's
      line is still stripped (the scan reads `updated_at`, which GitHub does not
      freeze), while the unedited case keeps the bot's line — `(ff-933)(n)` and
      its ordering control
- [x] Guard lineage intact: #875 hollow approvals, #876 in-place PATCH
      re-reviews, #894 all-decimal short SHA, #897 fence shapes, #917 UUID
      fragments — all 124 pre-existing assertions (the merge-base suite) still
      pass, none dropped or renamed (suite now 149/149)
- [x] Non-vacuity: every new direction case fails against main's evaluator
      (`PASS: 136  FAIL: 13`, all 13 failures direction cases) while every
      control — including the new `(ff-933)(n)` ordering control — passes on both
- [x] Live PR #929 replay: CodeAnt `counts_as_coverage` true → false;
      CodeRabbit byte-identical
- [x] Self-certifying check (this PR edits the evaluator merge-gate consumes):
      main's copy and this branch's produce **identical decision flags** on PR
      #951's own live payload, at all three pushed SHAs — so nothing here is
      graded by the guard it ships
- [x] CI-equivalent local run green: 65 bash suites, rule-lint, shellcheck

## Review rounds on this PR

Local CLI coverage stayed `none` (CodeAnt 403 twice — issue #643; the CodeRabbit
CLI's verification pass hung past the 2-minute bound and was dropped), and
CodeRabbit's GitHub review hit its rate limit, so the GitHub bots below were the
review signal. Two fix rounds, both in the **grant** direction — the direction
this PR exists to close, found in the fix itself:

- **`f7b4272`** — CodeAnt: the fence exemption was written as "keep any line that
  *starts with* a fence marker". Reproduced live: an echoed ` ```5acd1e2 ` scored
  `status_comment_names_head` and `counts_as_coverage` true off the author's
  words. Now truncated to the delimiter.
- **`861575c`** — BugBot: the echo scan read `created_at`, which GitHub freezes on
  an in-place edit — and in-place editing is routine for these bots (#876). A bot
  could edit an echo into a comment it opened *before* the author wrote the line
  and keep it. Scan now reads `updated_at`.

**Declined, with rationale on the thread:** CodeAnt also proposed normalising only
the Markdown quote prefix and preserving indentation in the echo key. Probed on
four shapes: it removes the false positive it describes (author indented,
reviewer flush-left) but stops catching an echo re-indented into a block, which
is the grant direction. Live PR #929 is flush-left on both sides, so neither
variant is needed for the reported trace; the tie is broken by this evaluator's
stated posture that over-matching withholds a merge (recoverable) while
under-matching grants one.

___

## **CodeAnt-AI Description**
Prevent quoted or echoed text from counting as review evidence

### What Changed
- Reviewers no longer receive HEAD credit when their comment only repeats a SHA-bearing line previously posted by someone outside the reviewers being evaluated.
- GitHub quote replies, copied text, and in-place edits are handled without removing a reviewer’s own status evidence or allowing later comments to rewrite history.
- Fenced code blocks remain structurally intact, while echoed fence metadata cannot smuggle SHA evidence.
- Added coverage for author echoes, self-quotation, ordering, reviewer-to-reviewer quotes, fence handling, and edited comments.
- Documented the evidence rules and their effects on review coverage and mismatch detection.

### Impact
`✅ Fewer false review approvals`
`✅ Accurate HEAD-reading evidence`
`✅ No self-report mismatches from quoted text`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

File: .claude/scripts/review-substance.sh [hunk 1]
```diff
@@ -238,6 +248,116 @@ OUT=$(printf '%s' "$INPUT" | jq -c \
...
+  # end-of-body and deletes rule-3 tokens that should have been admitted, which
+  # is the grant direction. A delimiter carries no commit id of its own, so
+  # keeping it costs nothing the echo rule was meant to buy. Pinned by
+  # (ff-933)(j); only PAIRED syntax needs this, which is why four-space-indented
+  # lines get no exception — dropping one removes its content too.
+  | def strip_echoed($ts):
+      [ ((. // "") | split("\n")[])
+        | . as $ln
+        | ($ln | norm_line) as $key
+        | (($key | startswith("```")) or ($key | startswith("~~~"))) as $is_fence
```
[the remark this answers is not in the document]
Comment (auerbachb) [reply .claude/scripts/review-substance.sh:372]: Fixed in `f7b4272`: confirmed and closed.

Reproduced first, on a live-shaped payload — author posts a line, bot echoes it, empty-body APPROVED on HEAD:

| echoed line | `names_head` | `counts_as_coverage` |
|---|---|---|
| ` ```5acd1e2 ` | true | **true** |
| `5acd1e2 is the head` | false | false |

So the exemption was readmitting precisely the token the echo filter had just refused — the #933 bug surviving through its own fix, on the grant side.

The fix takes your second option: an echoed fence line is truncated to its own delimiter run (leading whitespace and quote markers kept, info string and trailing text dropped) rather than exempted wholesale. Narrowing to "bare delimiters only" was the other option and it is not safe here — ` ```js ` is a real opener, and dropping it unbalances the #897 paired masking exactly the way case (j) exists to prevent. Truncation keeps the delimiter the masking regexes actually match on (they never read the info string) while removing the content.

Pinned by `(ff-933)(k)` (both spellings, asserted on `names_head` and `counts_as_coverage`) and `(ff-933)(l)` (a truncated opener still pairs with its closer and masks the block). Non-vacuity: those 4 (k) assertions fail against the pre-fix evaluator; 142 others pass on both sides. Suite 146/146, full CI discovery 65/65, shellcheck clean. The direction argument in the script header and `review-substance-evidence.md` were both updated — truncation only ever removes characters, so "tokens can only shrink" still holds.

File: .claude/scripts/review-substance.sh [hunk 2]
```diff
@@ -238,6 +248,116 @@ OUT=$(printf '%s' "$INPUT" | jq -c \
...
+  #    one channel (CodeRabbit CLI review of this PR — an earlier draft of this
+  #    comment wrongly listed it as withhold-only). Every other input pushes it
+  #    down: substantive can only fall, no_substantive_footprint can only
+  #    appear, and temporal_inversion and capability_failure are both suppressed
+  #    by $ext_substantive, which can only fall — so each can only newly fire.
+  #    The single upward path is a cleared self_report_mismatch on a reviewer
+  #    that is substantive on its own footprint, which is the corrected verdict,
+  #    not a loosened one.
+  | def norm_line:
+      sub("^[ \t]*(>[ \t]*)+"; "") | sub("^[ \t]+"; "") | sub("[ \t\r]+$"; "");
```
[the remark this answers is not in the document]
Comment (auerbachb) [reply .claude/scripts/review-substance.sh:321]: Verified against the code and the live payload, and declining this one — the change would move the guard in the grant direction.

Your read of the mechanism is right: `sub("^[ \t]+"; "")` makes the echo key indentation-insensitive, so an indented author line can suppress a reviewer line at another indent. I probed the proposed variant (`^[ \t]*(>[ \t]?)+` only, indentation preserved) against the current one on the four shapes that matter:

| shape | current | quote-prefix-only |
|---|---|---|
| live #929 echo (flush-left both sides) | caught | caught |
| GitHub "Quote reply" of an indented author line | caught | caught |
| echo re-indented into a block | caught | **missed** |
| author indented, reviewer flush-left (your FP) | suppressed | not suppressed |

The two variants differ on exactly two rows, and they are opposite directions. Preserving indentation buys the false positive you describe (a withheld merge, recoverable by re-review and push) and pays for it with row 3: an echo re-indented into a block stops being recognised, so the author's SHA manufactures HEAD evidence again — the grant this fix exists to close. The trade is irreducible; "same text, different indentation" is either an echo or it is not.

The tie-break is the script's stated posture, already load-bearing for rule 3 a few lines below: *"A false positive here withholds a merge (recoverable — re-review and push); the direction it structurally cannot take is the one that grants one."* Withholding wins.

Two things bound the exposure. The live PR #929 trace is flush-left on both sides, so neither variant is required for the reported case — the leading-whitespace strip is purely defensive. And the false positive needs the author to have already posted a line whose text is identical to the reviewer's modulo leading whitespace and case, with that line carrying the reviewer's only HEAD evidence; anything the reviewer says in its own words keeps its tokens (pinned by `(ff-933)(b)` and `(ff-933)(c)`).

Happy to revisit if a real payload shows this firing on a genuine reviewer line — that would be a concrete case to design against rather than a symmetric trade.

File: .claude/scripts/review-substance.sh [hunk 3]
```diff
@@ -360,7 +480,8 @@ OUT=$(printf '%s' "$INPUT" | jq -c \
     # ---- Conversation index: one pass, all regex work done here -------------
     ( [ $convo[]?
         | (.body // "") as $b
-        | (($b | sha_tokens)) as $tok
+        | (((.created_at // .updated_at) // "") | canon_ts) as $cts
+        | (($b | sha_tokens($cts))) as $tok
```
[the remark this answers is not in the document]
Comment (auerbachb) [reply .claude/scripts/review-substance.sh:523]: Fixed in `861575c`: valid, and the grant half of it is reachable in this repo rather than hypothetical.

Verified both halves against the code. The index (`echo_first`) keys on `created_at` and the scan passes `created_at` into `sha_tokens`/`strip_echoed`, and GitHub freezes `created_at` on an in-place edit. Editing in place is the norm for these two bots, not an edge case — CodeAnt PATCHes its review body on re-review (issue #876) and CodeRabbit rewrites its walkthrough comment — so a bot could edit an echo of the author's prose into a comment it had opened *before* the author wrote that line, sort ahead of the index entry, and keep it. HEAD evidence without the bot ever writing the SHA.

The fix takes `updated_at` on the scan side only, which leaves the two sides deliberately asymmetric but biased the same way — toward stripping:

- index → `created_at`, the **earliest** time a source line can have existed
- scan → `updated_at`, the **latest** time the scanned body can have been written

So an in-place edit can now only ever withhold coverage, never manufacture it, which matches the direction posture the rest of this evaluator is built on. Taking `updated_at` on the index side too would have been the wrong half: an unrelated typo fix on an author comment would push its `seen_at` past a genuine echo and re-open the same hole.

Unedited comments report `updated_at == created_at`, so the ordering guard — a later author comment cannot retroactively strip the bot — is unchanged. That is pinned explicitly by the unedited control in `(ff-933)(n)`, alongside the edited case; the control must stay true or the rule has quietly become "strip whenever the text matches". Non-vacuity: the 2 new `(n)` assertions fail against the pre-fix evaluator, the control passes on both sides. Suite 149/149, full CI discovery 65/65, shellcheck clean.

**Comment by auerbachb** [comment]
@cursor review
